### PR TITLE
issue1/re-org

### LINKED ISSUE #1: re-org
Reorganize the repo for easy installation.

### `TODO`

- [x] Move documentation to a `docs` folder
- [x] Move Jupyter notebooks to a `notebooks` folder
- [x] Move `*.gz` and `*.len` files to a `data` folder
- [x] Add a `conda` environment file
- [x] Add a `pyproject.toml` file
- [x] Create an installable package
- [x] Move script files to a `scripts` folder.
- [ ] Update the data loader to include new data checks
- [ ] Follow the `holoseq_display.py` script to create a visualization using available data

### PR DESCRIPTION
**Pull request by amaloney**
This PR does several things.

- Moves data to a dedicated `data` folder.
- Moves documentation to a dedicated `docs` folder.
- Moves documentation assets to a dedicated `docs/assets` folder.
- Moves Jupyter notebooks to a dedicated `notebooks` folder.
- Moves scripts into a dedicated `scripts` folder.
- Adds a `conda` environment file.
- Adds a `pyproject.toml` file.
- Adds pre-commit hooks
- Updates the `README.md` for formatting, syntax highlighting, and adds an installation section.
- Creates a `src` directory layout. This layout was chosen for future-proofing when we add tests to the repo.
- Runs `ruff` formatting and linting on all files, and adds a pre-commit hook for linting/formatting before commits.
- The example data for the squirrel was used to ensure the changes produced from the original scripts gave the final output shown in the README.
- Most original script files could not easily be fixed due to linting errors, however, as much was possible was done for this initial PR.
- Data loading memory usage was reduced by ~50%.
- Refinements in HoloViews could reduce memory usage further.

Resolves #1

### CODE REVIEW

**Review by fubar2 — APPROVED** [review]
As long as the repository meets your own internal standards and requirements it's good to go AFAIK. Have not tried to build it but good to see some reorganisation of the messy proof of concept - thanks @amaloney

File: README.md [hunk 1]
```diff
@@ -1,40 +1,53 @@
...
 
 ## Core idea: Features on intervals arranged along linear axes for browsing
 
-This proof of concept shows how millions of features can be generated along a couple of chromosomes and plotted using rasterize and datashader, with each tap converted into contig and offset. It runs in an IPython notebook, or if the dependencies are available, can be served from this repository's root, as:
+This proof of concept shows how millions of features can be generated along a couple of chromosomes
+and plotted using rasterize and datashader, with each tap converted into contig and offset. It runs
+in an IPython notebook, or if the dependencies are available, can be served from this repository's
+root, as:
 
 `panel serve holoSeq_random.py --show`
```
Comment (droumis) [inline README.md]:
```suggestion
`panel serve scripts/holoSeq_random.py --show`
```

File: environment.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+name: holoseq-dev
+
+channels:
+    - conda-forge
+
+dependencies:
+    - python >3.10,<3.13
+
+    # Package managers
+    - pip
```
Comment (droumis) [inline environment.yaml]:
```suggestion
name: holoseq-dev
channels:
  - conda-forge
  - pyviz/label/dev
dependencies:
  - python >3.10,<3.13
  - bokeh
  - dask[dataframe]
  - datashader
  - holoviews
  - pre-commit
  - ruff
  - jupyterlab
  - pytest
  - pytest-cov
  - nbsite>=0.8.4,<0.9.0
  - setuptools
  - setuptools-scm
```

File: README.md [hunk 3]
```diff
@@ -162,151 +183,235 @@ Expect to see output shown below, and a web browser window should pop open. Take
...
 
-Briefly, the framework creates the [minimum data required](https://github.com/fubar2/holoSeq/blob/main/HoloSeqOverview.md) to create a plot.
-A genome lengths file is required, and the named contigs can be reordered by name or length. The axes are defined by the ordering. The
-lengths are cumulated to give an offset to the first nucleotide of each contig, so the track can be read and feature locations 
-converted into the plot coordinate system, and stored as a compressed intermediate file. The display application reads these pre-computed plot 
-coordinate files, with enough metadata about the reference sequence to add tic marks to the axes and to back-calculate the stream of user tap coordinates. 
-A converter for PAF to compressed hseq format for input is available and was used to generate the demonstration. 
-Bigwig is working and other common genomic annotation formats, such as gff and vcf will follow. 
+Briefly, the framework creates the
+[minimum data required](https://github.com/fubar2/holoSeq/blob/main/HoloSeqOverview.md) to create a
```
Comment (droumis) [inline README.md]: update link

File: scripts/holoseq_display.py [hunk 4]
```diff
@@ -1,3 +1,5 @@
+# ruff: noqa
+
 # see https://github.com/fubar2/holoSeq
```
Comment (droumis) [inline scripts/holoseq_display.py]: update link

File: README.md [hunk 5]
```diff
@@ -1,40 +1,53 @@
...
 
 *Help wanted. PR and suggestions welcomed.*
 
-This is new work in progress. 
+This is new work in progress.
 
-Development started in late October 2024. 
-A draft framework [description and specification is here.](https://github.com/fubar2/holoSeq/blob/main/HoloSeqOverview.md). 
+Development started in late October 2024. A draft framework
+[description and specification is here.](https://github.com/fubar2/holoSeq/blob/main/HoloSeqOverview.md).
```
Comment (droumis) [inline README.md]: update link

File: README.md [hunk 6]
```diff
@@ -1,40 +1,53 @@
...
+internally consistent 2D space.
 
-This is proposed as a generalisable model for a linear display of genomic features located on a set of independent contigs in
-holoSeq displays.
+This is proposed as a generalisable model for a linear display of genomic features located on a set
+of independent contigs in holoSeq displays.
 
-```
+```python
 # see https://github.com/fubar2/holoSeq
```
Comment (droumis) [inline README.md]: update link

File: README.md [hunk 7]
```diff
@@ -122,38 +135,46 @@ pnc = pn.Column(
...
-Requires both Python and the python-venv module installed. 
-These steps ensure that the virtual environment directory can be made and deleted without any other changes to your system.
+Requires both Python and the python-venv module installed. These steps ensure that the virtual
+environment directory can be made and deleted without any other changes to your system.
 
 1. Clone this repository in a convenient and disposable directory:
 
-`git clone https://github.com/fubar2/holoSeq`
+```bash
+git clone https://github.com/fubar2/holoSeq
```
Comment (droumis) [inline README.md]: update link

File: README.md [hunk 8]
```diff
@@ -122,38 +135,46 @@ pnc = pn.Column(
...
 ```
 
-Expect to see output shown below, and a web browser window should pop open. Takes 10-20 seconds to read the 3.4M pairs and to show the interactive visualisation:
+Expect to see output shown below, and a web browser window should pop open. Takes 10-20 seconds to
+read the 3.4M pairs and to show the interactive visualisation:
 
-```
-(venv) ross@pn50:~/rossgit/holoSeq$ panel serve holoseq_display.py --args --inFile mUroPar1_cis1.hseq.gz --size 1000
+```bash
+panel serve holoseq_display.py --args --inFile mUroPar1_cis1.hseq.gz --size 1000
```
Comment (droumis) [inline README.md]: update path to script

File: README.md [hunk 9]
```diff
@@ -162,151 +183,235 @@ Expect to see output shown below, and a web browser window should pop open. Take
...
+marks to the axes and to back-calculate the stream of user tap coordinates. A converter for PAF to
+compressed hseq format for input is available and was used to generate the demonstration. Bigwig is
+working and other common genomic annotation formats, such as gff and vcf will follow.
 
 Multiple input files will produce a stack of plots that work independently:
 
-`panel serve holoseq_display.py --show --args --inFile mUroPar1.paf_cis1.hseq.gz small.paf_cis1.hseq.gz  --size 1000`
-
+```bash
+panel serve holoseq_display.py --show --args --inFile \
```
Comment (droumis) [inline README.md]: update path to script

File: README.md [hunk 10]
```diff
@@ -122,38 +135,46 @@ pnc = pn.Column(
...
+   and start the demonstration
 
-```
+```bash
 cd holoSeq
 python -m venv venv
 . venv/bin/activate
-pip install -r requirements.txt 
+pip install -r requirements.txt
 panel serve holoseq_display.py --show --args --inFile mUroPar1H1H2.paf_cisH1_hseq.gz --size 1000
```
Comment (droumis) [inline README.md]: update path to script

File: notebooks/hapsHiCpafHoloview.ipynb
Comment (droumis) [inline notebooks/hapsHiCpafHoloview.ipynb:1]: Although this notebook needs updates for install instructions and paths to data, let's leave it alone for now. It'll probably need @fubar2 to revise, given the expertise needed to provide appropriate narrative context as a sort of getting started guide.

File: environment.yaml (hunk 2 above)
Comment (droumis) [reply environment.yaml]: we can actually probably drop this file if we're going to generate the files in envs/ for the CI

**Review by droumis — APPROVED** [review]
The targeted performance improvements look good. Plenty left to do in future PRs for polishing and docs, tests, etc. I'll push a couple small additional changes to initiate some of that. In the future we can maybe also add some additional data validation.

Specifically, we need to eventually update all the paths to scripts and data files and correct any references to the former gh org, but that's beyond scope of this PR.

**Comment by droumis** [comment]
@amaloney can you please address the CI issues and then merge this PR

File: environment.yaml (hunk 2 above)
Comment (amaloney) [reply environment.yaml]: I'm okay dropping this, and adding a reference to it in the `README.md` file on how to create a conda environment. We can even add a section about how to use `pyproject2conda` in order to get an `environment.yaml` file so a user can use conda exclusively.

What I'm not sure about is the maintenance burden of having several environment files in the repo due to CI. The CI can be modified to not need these environments, so I'm inclined to remove them.

Comment (droumis) [reply environment.yaml]: I'm fine with all of that

**Pull request merged**